### PR TITLE
Changes for Personalized Page Rank : Jira:1084

### PR DESCRIPTION
**Pull request by hpandeycodeit**
Jira : 1084
This PR contains changes for Personalized Page Rank.

-  Added extra parameter, nodes_of_interest in main pagerank function. 
-  Added a new Function get_query_params_for_ppr  in pagerank.py_in to calculate random_jump_probabilty based on the user provided input nodes. 
-  Added a condition, when the user provided nodes are present then Personalized Page Rank will be executed otherwise regular Page Rank will run. 
-  Added an example function in pagerank.sql_in
- The extra parameter nodes_of_interest is also added in the calling functions in pagerank.sql_in

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/379/

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 1]
```diff
@@ -44,29 +44,40 @@ from utilities.utilities import add_postfix
...
-                                            threshold, max_iter,
-                                            edge_table, grouping_cols_list)
+                                      threshold, max_iter,
+                                      edge_table, grouping_cols_list)
     _assert(damping_factor >= 0.0 and damping_factor <= 1.0,
             "PageRank: Invalid damping factor value ({0}), must be between 0 and 1.".
             format(damping_factor))
 
+    if nodes_of_interest:
+        vertices = plpy.execute("""
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.py_in]: Variable name can be num_vertices since it is a count number

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 2]
```diff
@@ -44,29 +44,40 @@ from utilities.utilities import add_postfix
...
+                   SELECT count({vertex_id}) AS cnt FROM {vertex_table}
+                   WHERE {vertex_id} = ANY(ARRAY{nodes_of_interest})
+                   """.format(**locals()))[0]["cnt"]
+        if vertices != len(nodes_of_interest):
+            plpy.error("PageRank: Invalid value for {0}, must be a subset of the vertex_table".format(
+                nodes_of_interest))
+
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.py_in]: Besides checking total count, what if there is any invalid vertex id in nodes_of_interest?

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 3]
```diff
@@ -149,25 +164,39 @@ def pagerank(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
...
+                                              personalized_nodes)
+        total_ppr_nodes = ppr_params[0]
+        random_jump_prob_ppr = ppr_params[1]
+        ppr_join_clause = ppr_params[2]
+
         random_probability = (1.0 - damping_factor) / n_vertices
+        if total_ppr_nodes > 0:
+            random_jump_prob = random_jump_prob_ppr
+        else:
+            random_jump_prob = random_probability
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.py_in]: Can move (1.0 - damping_factor) / n_vertices here since random_probability is not used anywhere else.

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 4]
```diff
@@ -527,14 +562,63 @@ def pagerank(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
...
+         @param edge_temp_table
+         @param vertex_table
+         @param cur_distribution
+
+     Returns :
+             (Integer, String, String)
+
+    """
+    total_ppr_nodes = 0
+    random_jump_prob_ppr = ''
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.py_in]: This is more like a sql condition/clause, not a number of prob, should we rename it somehow?

File: src/ports/postgres/modules/graph/test/pagerank.sql_in [hunk 5]
```diff
@@ -66,7 +66,12 @@ SELECT pagerank(
              'id',            -- Vertix id column
              '"EDGE"',          -- "EDGE" table
              'src=src, dest=dest', -- "EDGE" args
-             'pagerank_out'); -- Output table of PageRank
+             'pagerank_out',
+              NULL,
+              NULL,
+              NULL,
+              NULL,
+             NULL); -- Output table of PageRank
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/test/pagerank.sql_in]: Move comment '-- Output table of PageRank' to upper column.

File: src/ports/postgres/modules/graph/test/pagerank.sql_in [hunk 6]
```diff
@@ -84,7 +89,8 @@ SELECT pagerank(
              NULL,
              NULL,
              NULL,
-             'user_id');
+             'user_id',
+             NULL);
 
 -- View the PageRank of all vertices, sorted by their scores.
 SELECT assert(relative_error(SUM(pagerank), 1) < 0.00001,
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/test/pagerank.sql_in]: We may need at least two test cases with nodes_of_interest not null and with/without grouping.

File: src/ports/postgres/modules/graph/pagerank.sql_in [hunk 7]
```diff
@@ -120,6 +121,10 @@ distribution per group. When this value is NULL, no grouping is used and
 a single model is generated for all data.
 @note Expressions are not currently supported for 'grouping_cols'.</dd>
 
+<dt> nodes_of_interest (optional) </dt>
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.sql_in]: @fmcquillan99 Do we also need additional explanation of personalized pagerank somewhere in the user doc?

Comment (fmcquillan99) [reply src/ports/postgres/modules/graph/pagerank.sql_in]: Yes user doc will need more explanation of what is personalized page rank and how it works and examples.

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 8]
```diff
@@ -527,14 +562,63 @@ def pagerank(schema_madlib, vertex_table, vertex_id, edge_table, edge_args,
...
+        # In case of PPR, Assign the Random jump probability to the nodes_of_interest only.
+        # For rest of the nodes, Random jump probability  will be zero.
+        random_jump_prob_ppr = """ CASE when {edge_temp_table}.dest = ANY(ARRAY{nodes_of_interest})
+                                THEN {prob_value}
+                                ELSE 0
+                                END """.format(**locals())
+    return(total_ppr_nodes, random_jump_prob_ppr, ppr_join_clause)
+
+
 def pagerank_help(schema_madlib, message, **kwargs):
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.py_in]: Added the explanation and example in the helper function

File: src/ports/postgres/modules/graph/pagerank.py_in (hunk 1 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.py_in]: Changed vertices to vertices_count as discussed.

File: src/ports/postgres/modules/graph/pagerank.py_in (hunk 4 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.py_in]: renamed this variable to ppr_random_prob_clause

File: src/ports/postgres/modules/graph/test/pagerank.sql_in (hunk 5 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/test/pagerank.sql_in]: This is corrected as well.

File: src/ports/postgres/modules/graph/pagerank.sql_in [hunk 9]
```diff
@@ -273,6 +278,48 @@ SELECT * FROM pagerank_out_summary ORDER BY user_id;
...
+                       'vertex',             -- Vertex table
+                       'id',                 -- Vertix id column
+                       'edge',               -- Edge table
+                       'src=src, dest=dest', -- Comma delimted string of edge arguments
+                       'pagerank_out',       -- Output table of PageRank 
+                        NULL,                -- Default damping factor (0.85)
+                        NULL,                -- Default max iters (100)
+                        NULL,                -- Default Threshold 
+                        NULL,                -- No Grouping 
+                       '{2,4}');             -- Personlized Nodes
```
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.sql_in]: Added this format in helper function example and also added the comment in the example in above file.

File: src/ports/postgres/modules/graph/pagerank.py_in (hunk 3 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.py_in]: This is getting used in when we create nodes_with_no_incoming_edges table.

File: src/ports/postgres/modules/graph/pagerank.sql_in (hunk 7 above)
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.sql_in]: I will add more explanation to the docs on how personalized page rank works. And still working on Test cases.

File: src/ports/postgres/modules/graph/pagerank.py_in (hunk 8 above)
[the remark this answers is not in the document]
Comment (hpandeycodeit) [reply src/ports/postgres/modules/graph/pagerank.py_in]: Also changed the  DISTRIBUTED BY clause in the get_query_params_for_ppr for the above error.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/387/

**Comment by kaknikhil** [comment]
@hpandeycodeit 
Can you add more description in the final commit (maybe use the same description as the PR).
This is the convention we use for our commit messages.
```
Commit Title

JIRA: MADLIB-1084

Detailed description
```

See more here : https://cwiki.apache.org/confluence/display/MADLIB/Contribution+Guidelines

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/404/

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 10]
```diff
@@ -551,14 +680,16 @@ def pagerank_help(schema_madlib, message, **kwargs):
...
                                      -- (DEFAULT = 0.85)
     max_iter      INTEGER, -- Maximum iteration number (DEFAULT = 100)
     threshold     DOUBLE PRECISION, -- Stopping criteria (DEFAULT = 1/(N*1000),
                                     -- N is number of vertices in the graph)
-    grouping_col  TEXT -- Comma separated column names to group on
+    grouping_col  TEXT, -- Comma separated column names to group on
                        -- (DEFAULT = NULL, no grouping)
+    nodes_of_interest ARRAY OF INTEGER -- A comma seperated list of vertices
+                                          or nodes for personalized page rank.
 """) + """
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.py_in]: indent left side, and indent comment(--) right

File: src/ports/postgres/modules/graph/pagerank.py_in [hunk 11]
```diff
@@ -647,6 +778,26 @@ SELECT * FROM pagerank_out ORDER BY user_id, pagerank DESC;
...
+SELECT madlib.pagerank(
+                       'vertex',             -- Vertex table
+                       'id',                 -- Vertix id column
+                       'edge',               -- Edge table
+                       'src=src, dest=dest', -- Comma delimted string of edge arguments
+                       'pagerank_out',       -- Output table of PageRank
+                        NULL,                -- Default damping factor (0.85)
+                        NULL,                -- Default max iters (100)
+                        NULL,                -- Default Threshold
+                        NULL,                -- No Grouping
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/pagerank.py_in]: move those NULLs one space left

File: src/ports/postgres/modules/graph/test/pagerank.sql_in [hunk 12]
```diff
@@ -66,7 +66,12 @@ SELECT pagerank(
              'id',            -- Vertix id column
              '"EDGE"',          -- "EDGE" table
              'src=src, dest=dest', -- "EDGE" args
-             'pagerank_out'); -- Output table of PageRank
+             'pagerank_out',    -- Output table of PageRank
+              NULL,             -- Default damping factor (0.85)
+              NULL,             -- Default max iters (100)
+              NULL,             -- Default Threshold 
+              NULL,             -- No Grouping 
+             NULL);             -- Personlized Nodes
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/test/pagerank.sql_in]: In this case, we can remove the last 5 NULLs since they are all optional.

File: src/ports/postgres/modules/graph/test/pagerank.sql_in [hunk 13]
```diff
@@ -95,6 +101,49 @@ SELECT assert(relative_error(SUM(pagerank), 1) < 0.00001,
...
+             'pagerank_ppr_out', -- Output table of PageRank
+             NULL,              -- Default damping factor (0.85)
+             NULL,              -- Default max iters (100)
+             NULL,              -- Default Threshold 
+             NULL,         -- Grouping column
+            '{1,3}');             -- Personlized Nodes
+
+
+-- View the PageRank of all vertices, sorted by their scores.
+SELECT assert(relative_error(SUM(pagerank), 1) < 0.00124,
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/test/pagerank.sql_in]: Is this  0.00124 based on current test result? Can we make it smaller?

File: src/ports/postgres/modules/graph/test/pagerank.sql_in [hunk 14]
```diff
@@ -95,6 +101,49 @@ SELECT assert(relative_error(SUM(pagerank), 1) < 0.00001,
...
+             '"EDGE"',          -- "EDGE" table
+             'src=src, dest=dest', -- "EDGE" args
+             'pagerank_ppr_grp_out', -- Output table of PageRank
+             NULL,              -- Default damping factor (0.85)
+             NULL,              -- Default max iters (100)
+             NULL,              -- Default Threshold 
+             'user_id',         -- Grouping column
+            '{1,3}');             -- Personlized Nodes
+
+SELECT assert(count(*) = 14, 'Tuple count for Pagerank out table != 14') FROM pagerank_ppr_grp_out;
```
Comment (jingyimei) [inline src/ports/postgres/modules/graph/test/pagerank.sql_in]: can we do similar assertion here by group?

File: src/ports/postgres/modules/graph/pagerank.py_in (hunk 3 above)
Comment (jingyimei) [reply src/ports/postgres/modules/graph/pagerank.py_in]: Got it.

File: src/ports/postgres/modules/graph/pagerank.sql_in (hunk 9 above)
[the remark this answers is not in the document]
Comment (jingyimei) [reply src/ports/postgres/modules/graph/pagerank.sql_in]: Great

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/419/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/426/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/427/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/434/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/435/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/437/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/438/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/440/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/441/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/442/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/446/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/447/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/451/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/452/
